File: backend/src/scrapers/bestbuy.ts
/**
 * Best Buy Products API adapter
 * Docs: https://bestbuyapis.github.io/api-documentation/
 */

const CATEGORY_MAP: Record<string, string> = {
  cpu: 'abcat0507002',
  gpu: 'abcat0507002',
  motherboard: 'abcat0507006',
  ram: 'abcat0503002',
  storage: 'pcmcat158900050001',
  psu: 'abcat0507009',
  case: 'abcat0507008',
  cooler: 'abcat0507005',
};

const SPEC_MAP: Record<string, string> = {
  'Number of Cores': 'cores',
  'Processor Speed': 'base_clock',
  'Socket Type': 'socket',
  'Memory Type': 'memory_type',
  'Memory Speed': 'speed',
  'Total Memory Capacity': 'capacity',
  'Graphics Memory': 'vram',
  'Form Factor': 'form_factor',
  'Wattage': 'wattage',
  'Chipset Brand': 'chipset',
};

export interface BestBuyProduct {
  sku: string;
  name: string;
  manufacturer: string;
  price: number;
  regularPrice: number;
  onSale: boolean;
  url: string;
  inStock: boolean;
  specs: Record<string, string>;
  categoryPath: string[];
}

interface RawProduct {
  sku: number;
  name: string;
  manufacturer: string;
  salePrice: number;
  regularPrice: number;
  onSale: boolean;
  url: string;
  categoryPath: Array<{ id: string; name: string }>;
  details: Array<{ name: string; value: string }>;
  inStoreAvailability: boolean;
  onlineAvailability: boolean;
}

interface ApiResponse {
  products: RawProduct[];
  total: number;
  from: number;
  to: number;
}

export class BestBuyAdapter {
  private apiKey: string;
  private baseUrl = 'https://api.bestbuy.com/v1';
  private requestsPerSecond: number;
  private lastRequestTime = 0;

  constructor(apiKey: string, requestsPerSecond = 5) {
    this.apiKey = apiKey;
    this.requestsPerSecond = requestsPerSecond;
  }

  getCategoryId(category: string): string {
    return CATEGORY_MAP[category] || category;
  }

  private async makeRequest(endpoint: string, params: Record<string, string> = {}): Promise<ApiResponse> {
    // Rate limiting
    const now = Date.now();
    const minInterval = 1000 / this.requestsPerSecond;
    const elapsed = now - this.lastRequestTime;
    if (elapsed < minInterval) {
      await new Promise(resolve => setTimeout(resolve, minInterval - elapsed));
    }

    const searchParams = new URLSearchParams({
      ...params,
      apiKey: this.apiKey,
      format: 'json',
    });

    const url = `${this.baseUrl}/${endpoint}?${searchParams}`;
    
    const response = await fetch(url);
    this.lastRequestTime = Date.now();

    if (!response.ok) {
      if (response.status === 429) {
        // Rate limited - wait and retry
        await new Promise(resolve => setTimeout(resolve, 5000));
        return this.makeRequest(endpoint, params);
      }
      throw new Error(`API error: ${response.status} ${response.statusText}`);
    }

    return response.json();
  }

  private parseProduct(raw: RawProduct): BestBuyProduct {
    const specs: Record<string, string> = {};
    
    for (const detail of raw.details || []) {
      const mappedKey = SPEC_MAP[detail.name];
      if (mappedKey) {
        // Clean up values (remove units)
        const cleanValue = detail.value.split(' ')[0];
        specs[mappedKey] = cleanValue;
      }
    }

    return {
      sku: String(raw.sku),
      name: raw.name,
      manufacturer: raw.manufacturer,
      price: raw.salePrice,
      regularPrice: raw.regularPrice,
      onSale: raw.onSale,
      url: raw.url,
      inStock: raw.onlineAvailability || raw.inStoreAvailability,
      specs,
      categoryPath: (raw.categoryPath || []).map(c => c.name),
    };
  }

  async fetchCategory(
    category: string,
    pageSize = 100,
    maxPages = 10
  ): Promise<BestBuyProduct[]> {
    const categoryId = this.getCategoryId(category);
    const products: BestBuyProduct[] = [];
    let page = 1;

    while (page <= maxPages) {
      const endpoint = `products(categoryPath.id=${categoryId})`;
      const params = {
        show: 'sku,name,manufacturer,salePrice,regularPrice,onSale,url,categoryPath,details,inStoreAvailability,onlineAvailability',
        pageSize: String(Math.min(pageSize, 100)),
        page: String(page),
      };

      const data = await this.makeRequest(endpoint, params);
      
      if (!data.products?.length) break;

      for (const raw of data.products) {
        products.push(this.parseProduct(raw));
      }

      if (data.to >= data.total) break;
      page++;
    }

    return products;
  }

  async search(query: string, limit = 20): Promise<BestBuyProduct[]> {
    const endpoint = `products(search=${encodeURIComponent(query)})`;
    const params = {
      show: 'sku,name,manufacturer,salePrice,regularPrice,onSale,url,categoryPath,details,inStoreAvailability,onlineAvailability',
      pageSize: String(Math.min(limit, 100)),
    };

    const data = await this.makeRequest(endpoint, params);
    return (data.products || []).map(p => this.parseProduct(p));
  }

  async getProduct(sku: string): Promise<BestBuyProduct | null> {
    const endpoint = `products(sku=${sku})`;
    const params = {
      show: 'sku,name,manufacturer,salePrice,regularPrice,onSale,url,categoryPath,details,inStoreAvailability,onlineAvailability',
    };

    const data = await this.makeRequest(endpoint, params);
    if (!data.products?.length) return null;

    return this.parseProduct(data.products[0]);
  }
}

File: backend/app/scrapers/bestbuy.py
"""
Best Buy Products API adapter.

Official API documentation: https://bestbuyapis.github.io/api-documentation/
"""

import asyncio
from decimal import Decimal
from typing import Any, Optional
import httpx


# Map our categories to Best Buy category IDs
CATEGORY_MAP = {
    "cpu": "abcat0507002",           # CPUs / Processors
    "gpu": "abcat0507002",           # Graphics Cards (note: may need refinement)
    "motherboard": "abcat0507006",   # Motherboards
    "ram": "abcat0503002",           # Computer Memory
    "storage": "pcmcat158900050001", # Solid State Drives
    "psu": "abcat0507009",           # Power Supplies
    "case": "abcat0507008",          # Computer Cases
    "cooler": "abcat0507005",        # CPU Cooling
}

# Map Best Buy spec names to our normalized keys
SPEC_MAP = {
    "Number of Cores": "cores",
    "Processor Speed": "base_clock",
    "Socket Type": "socket",
    "Memory Type": "memory_type",
    "Memory Speed": "speed",
    "Total Memory Capacity": "capacity",
    "Graphics Memory": "vram",
    "Form Factor": "form_factor",
    "Wattage": "wattage",
    "Chipset Brand": "chipset",
}


class BestBuyAdapter:
    """
    Adapter for Best Buy's Products API.
    
    Usage:
        adapter = BestBuyAdapter(api_key="YOUR_KEY")
        products = await adapter.fetch_category("cpu")
    """
    
    BASE_URL = "https://api.bestbuy.com/v1"
    
    def __init__(
        self,
        api_key: str,
        requests_per_second: int = 5,
        timeout: float = 30.0,
    ):
        self.api_key = api_key
        self.requests_per_second = requests_per_second
        self.timeout = timeout
        self._last_request_time = 0.0
        self._client: Optional[httpx.AsyncClient] = None
    
    async def _get_client(self) -> httpx.AsyncClient:
        """Get or create HTTP client."""
        if self._client is None:
            self._client = httpx.AsyncClient(timeout=self.timeout)
        return self._client
    
    async def _make_request(self, endpoint: str, params: Optional[dict] = None) -> dict:
        """Make a rate-limited request to the API."""
        # Rate limiting
        now = asyncio.get_event_loop().time()
        min_interval = 1.0 / self.requests_per_second
        elapsed = now - self._last_request_time
        if elapsed < min_interval:
            await asyncio.sleep(min_interval - elapsed)
        
        client = await self._get_client()
        
        # Add API key
        params = params or {}
        params["apiKey"] = self.api_key
        params["format"] = "json"
        
        url = f"{self.BASE_URL}/{endpoint}"
        
        try:
            response = await client.get(url, params=params)
            response.raise_for_status()
            self._last_request_time = asyncio.get_event_loop().time()
            return response.json()
        except httpx.HTTPStatusError as e:
            if e.response.status_code == 429:
                # Rate limited - wait and retry
                await asyncio.sleep(5)
                response = await client.get(url, params=params)
                response.raise_for_status()
                return response.json()
            raise
    
    def get_category_id(self, category: str) -> str:
        """Get Best Buy category ID from our category name."""
        return CATEGORY_MAP.get(category, category)
    
    def _parse_product(self, raw: dict) -> dict:
        """Parse a raw Best Buy product into our format."""
        # Extract specs from details array
        specs = {}
        for detail in raw.get("details", []):
            name = detail.get("name", "")
            value = detail.get("value", "")
            
            # Map to our normalized keys
            if name in SPEC_MAP:
                # Clean up values (remove units like "gigahertz")
                clean_value = value.split()[0] if value else ""
                specs[SPEC_MAP[name]] = clean_value
        
        # Determine stock status
        in_stock = raw.get("onlineAvailability", False) or raw.get("inStoreAvailability", False)
        
        return {
            "sku": str(raw.get("sku", "")),
            "name": raw.get("name", ""),
            "manufacturer": raw.get("manufacturer", ""),
            "price": Decimal(str(raw.get("salePrice", 0))),
            "regular_price": Decimal(str(raw.get("regularPrice", 0))),
            "on_sale": raw.get("onSale", False),
            "url": raw.get("url", ""),
            "in_stock": in_stock,
            "specs": specs,
            "category_path": [c.get("name") for c in raw.get("categoryPath", [])],
        }
    
    async def fetch_category(
        self,
        category: str,
        page_size: int = 100,
        max_pages: int = 10,
    ) -> list[dict]:
        """
        Fetch all products in a category.
        
        Args:
            category: Our category name (cpu, gpu, etc.)
            page_size: Products per page (max 100)
            max_pages: Maximum pages to fetch
        
        Returns:
            List of parsed products
        """
        category_id = self.get_category_id(category)
        products = []
        page = 1
        
        while page <= max_pages:
            # Build query
            endpoint = f"products(categoryPath.id={category_id})"
            params = {
                "show": "sku,name,manufacturer,salePrice,regularPrice,onSale,url,categoryPath,details,inStoreAvailability,onlineAvailability",
                "pageSize": min(page_size, 100),
                "page": page,
            }
            
            data = await self._make_request(endpoint, params)
            
            raw_products = data.get("products", [])
            if not raw_products:
                break
            
            for raw in raw_products:
                products.append(self._parse_product(raw))
            
            # Check if there are more pages
            total = data.get("total", 0)
            fetched = data.get("to", 0)
            if fetched >= total:
                break
            
            page += 1
        
        return products
    
    async def search(self, query: str, limit: int = 20) -> list[dict]:
        """
        Search for products by keyword.
        
        Args:
            query: Search terms
            limit: Max results to return
        
        Returns:
            List of parsed products
        """
        # URL-encode and build search endpoint
        endpoint = f"products(search={query})"
        params = {
            "show": "sku,name,manufacturer,salePrice,regularPrice,onSale,url,categoryPath,details,inStoreAvailability,onlineAvailability",
            "pageSize": min(limit, 100),
        }
        
        data = await self._make_request(endpoint, params)
        
        return [self._parse_product(raw) for raw in data.get("products", [])]
    
    async def get_product(self, sku: str) -> Optional[dict]:
        """
        Get a specific product by SKU.
        
        Args:
            sku: Best Buy SKU
        
        Returns:
            Parsed product or None if not found
        """
        endpoint = f"products(sku={sku})"
        params = {
            "show": "sku,name,manufacturer,salePrice,regularPrice,onSale,url,categoryPath,details,inStoreAvailability,onlineAvailability",
        }
        
        data = await self._make_request(endpoint, params)
        
        products = data.get("products", [])
        if not products:
            return None
        
        return self._parse_product(products[0])
    
    async def close(self):
        """Close the HTTP client."""
        if self._client:
            await self._client.aclose()
            self._client = None
